src/plugins/embeddable/public/lib/embeddable_saved_object_registry/embeddable_saved_object_registry.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { PresentationContainer } from '@kbn/presentation-containers';
+import { FinderAttributes, SavedObjectCommon } from '@kbn/saved-objects-finder-plugin/common';
+import { SavedObjectMetaData } from '@kbn/saved-objects-finder-plugin/public';
+
+type SOToEmbeddable<TSavedObjectAttributes extends FinderAttributes = FinderAttributes> = (
+  container: PresentationContainer,
+  savedObject: SavedObjectCommon<TSavedObjectAttributes>
+) => void;
+
+export type ReactEmbeddableSavedObject<
+  TSavedObjectAttributes extends FinderAttributes = FinderAttributes
+> = {
+  onAdd: SOToEmbeddable<TSavedObjectAttributes>;
+  savedObjectMetaData: SavedObjectMetaData;
```

Comment:
Is it necessary to nest `SavedObjectMetaData` properties under `savedObjectMetaData` key?

`AddPanelFlyout` filters out factories without a type so I think type should be required. 

`registerReactEmbeddableSavedObject` takes a `type` parameter and then `savedObjectMetaData`. What is different about `type` and `savedObjectMetaData.type` when calling `registerReactEmbeddableSavedObject`?

How about just typing as
```
{
  onAdd,
  type,
  displayName,
  getIcon,
}
```

Reply:
> What is different about type and savedObjectMetaData.type when calling registerReactEmbeddableSavedObject?

`type` refers to the embeddable type that is passed to [https://github.com/elastic/kibana/blob/main/src/plugins/embeddable/public/react_embeddable_system/react_embeddable_registry.ts#L17](https://github.com/elastic/kibana/blob/main/src/plugins/embeddable/public/react_embeddable_system/react_embeddable_registry.ts#L17) `savedObjectMetaData.type` is the content type that is stored with the saved object.

In most cases, our legacy embeddables use the same constant for registering the embeddable factory type and the saved object type. [Maps](https://github.com/elastic/kibana/blob/main/x-pack/plugins/maps/public/embeddable/map_embeddable_factory.ts#L17-L24) and [Lens](https://github.com/elastic/kibana/blob/main/x-pack/plugins/lens/public/embeddable/embeddable_factory.ts#L59-L64) are good examples of this.

Some other plugins use separate constants for registering their embeddable factories and saved objects. Visualization uses the [`VISUALIZE_EMBEDDABLE_TYPE` constant](https://github.com/elastic/kibana/blob/main/src/plugins/visualizations/public/embeddable/visualize_embeddable_factory.tsx#L82) to define the embeddable factory type, but [uses `"visualization"`](https://github.com/elastic/kibana/blob/main/src/plugins/visualizations/public/embeddable/visualize_embeddable_factory.tsx#L93) to define the saved object type. But it should be noted that the `VISUALIZE_EMBEDDABLE_TYPE` is also equal to `visualization`. I'm not sure if there was any reason to make these distinct constants. 

The [Search embeddable factory](https://github.com/elastic/kibana/blob/main/src/plugins/discover/public/embeddable/search_embeddable_factory.ts#L31-L36) also uses distinct, but otherwise equal constants for the embeddable factory type and the saved object type. 

Perhaps as part of this refactoring, we can unify the embeddable type with the saved object type (for plugins that have saved objects). WDYT @ThomThomson?

> Is it necessary to nest SavedObjectMetaData properties under savedObjectMetaData key?
> How about just typing as

> {
  onAdd,
  type,
  displayName,
  getIcon,
}

I think we still need to include the optional `savedObjectMetaData.getSavedObjectSubType` method so we can report accurate telemetry for visualize objects. Or we could encourage embeddables to report their own telemetry in the `onAdd` method rather than doing so in the flyout.

Reply:
I think we still need to support a separation between `savedObjectType` and `embeddableType`. Those are fundamentally two different concepts. Think of this registry as a mapping table of SOs to embeddables.

Reply:
> I think we still need to support a separation between savedObjectType and embeddableType. Those are fundamentally two different concepts. Think of this registry as a mapping table of SOs to embeddables.

Makes sense. I spoke offline with @nreese about this and we agreed on a solution that distinctly identifies these types in the registry.

Reply:
@nreese 

> Is it necessary to nest SavedObjectMetaData properties under savedObjectMetaData key?

I can flatten the type, but I still [need to pass an object matching `SavedObjectMetaData` to the `SavedObjectFinder`](https://github.com/elastic/kibana/pull/179289/files#diff-9de3e83df231f47b3e03a18a65b98de136ea9545a386fecc75c7b2cad9def170R143). If flattening the type helps with developer experience, then I guess I could write a function that maps the flattened type to the `SavedObjectMetaData` type. WDYT?

Reply:
maybe just throw if SavedObjectMetaData.type is not provided then so there is immediate feedback that the registry entry is a failure

Reply:
Oops, sorry. I misunderstood what you were proposing. I think your original proposal will work. I'm updating tests and will finish the PR soon.